scheduler/scheduler.go

```diff
@@ -227,30 +233,117 @@ func (driver *MesosSchedulerDriver) masterDetected(master *mesos.MasterInfo) {
 		}
 
 		driver.MasterPid = pid // save for downstream ops.
+		driver.tryAuthentication()
+	} else {
+		log.Infoln("No master detected.")
+	}
+}
+
+func (driver *MesosSchedulerDriver) tryAuthentication() {
+	if driver.authenticated {
+		// programming error
+		panic("already authenticated")
+	}
+
+	masterPid := driver.MasterPid // save for referencing later in goroutine
+	if masterPid == nil {
+		log.Info("skipping authentication attempt because we lost the master")
+		return
+	}
+
+	if driver.authenticating != nil {
```

Comment:
Another way to model this would be to have an `authenticating bool` and an `authenticationComplete chan struct{}`. I was basically trying to combine the two into a single thing. When `authenticating != nil` then there's an authentication in progress that can be waited for by `<-authenticating`.

Chan's are really great for async communication. They're a little different from promises and futures though and require extra work to avoid pitfalls like panics on `close(someChan)` when that chan is already closed. Really what I think I wanted was a future that I could discard when I was done with it - potentially multiple times without error. I've actually been toying with some ideas for a promise/future library in go (https://github.com/jdef/sync/tree/master/pkg/future) but it doesn't have much testing and I didn't want to bring in a new lib just for this.